src/main/java/org/opensrp/repository/postgres/IdentifierSourceRepositoryImpl.java

```diff
@@ -0,0 +1,166 @@
+import org.opensrp.util.IdentifierValidatorAlgorithm;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Repository("identifierSourceRepositoryPostgres")
+public class IdentifierSourceRepositoryImpl extends BaseRepositoryImpl<IdentifierSource>
+		implements IdentifierSourceRepository {
+
+	@Autowired
+	private CustomIdentifierSourceMapper customIdentifierSourceMapper;
+
+	@Override
+	public IdentifierSource findByIdentifier(String identifier) {
+		org.opensrp.domain.postgres.IdentifierSource identifierSource = customIdentifierSourceMapper
+				.selectByIdentifier(identifier);
+		return convert(identifierSource);
+	}
+
+	@Override
+	public IdentifierSource get(String id) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	public void add(IdentifierSource entity) {
+		if (entity == null || entity.getIdentifier() == null) {
+			return;
+		}
+
+		if (retrievePrimaryKey(entity) != null) { // IdentifierSource already added
+			return;
+		}
+		org.opensrp.domain.postgres.IdentifierSource pgIdentifierSource = convert(entity, null);
+		if (pgIdentifierSource == null) {
+			return;
+		}
+
+		int rowsAffected = customIdentifierSourceMapper.insertSelectiveAndSetId(pgIdentifierSource);
+		if (rowsAffected < 1 || pgIdentifierSource.getId() == null) {
+			return;
+		}
+
+	}
+
+	@Override
+	public void update(IdentifierSource entity) {
+		if (entity == null || entity.getId() == null || entity.getIdentifier() == null) {
+			return;
+		}
+		Long id = retrievePrimaryKey(entity);
+
+		if (id == null) { //Identifier Source doesn't not exist
+			return;
+		}
+
+		org.opensrp.domain.postgres.IdentifierSource pgIdentifierSource = convert(entity, id);
+		customIdentifierSourceMapper.updateByPrimaryKey(pgIdentifierSource);
+	}
+
+	@Override
+	public List<IdentifierSource> getAll() {
+		IdentifierSourceExample identifierSourceExample = new IdentifierSourceExample();
+		identifierSourceExample.createCriteria().andIdentifierIsNotNull();
+		List<org.opensrp.domain.postgres.IdentifierSource> identifierSources = customIdentifierSourceMapper
+				.selectMany(identifierSourceExample, 0, DEFAULT_FETCH_SIZE);
+		return convert(identifierSources);
+	}
+
+	@Override
+	public void safeRemove(IdentifierSource entity) {
+		throw new NotImplementedException();
+	}
+
+	@Override
+	protected Long retrievePrimaryKey(IdentifierSource entity) {
+
+		IdentifierSourceExample identifierSourceExample = new IdentifierSourceExample();
+		IdentifierSourceExample.Criteria criteria = identifierSourceExample.createCriteria();
+		criteria.andIdentifierIsNotNull();
+
+		criteria.andIdEqualTo(entity.getId() != null ? entity.getId() : 0);
+		org.opensrp.domain.postgres.IdentifierSource identifierSource = customIdentifierSourceMapper
+				.selectOne(identifierSourceExample);
+		if (identifierSource == null) {
+			return null;
+		}
+		return identifierSource.getId();
```

Comment:
There is a bug on this method.On the criteria this is not required `criteria.andIdentifierIsNotNull();`

since id can be null you should use identifier to search. 

if you want want to use id use only if its not null and its not zero, otherwise use identifier

Reply:
Checking with id first if it is not null and non-zero, otherwise checking with Identifier.